### ISSUE TITLE
Grape and Protecting Against Mass Assignment Abuse

### ISSUE THREAD

**Issue by jonallured** [ruby-grape/grape#2416]
tl;dr: it's too easy to write Grape endpoints that are not protected against [Mass Assignment abuse](https://en.wikipedia.org/wiki/Mass_assignment_vulnerability).

I've created [an example Rails application called `PretendGravity`](https://github.com/jonallured/pretend_gravity), an homage to the main API server at Artsy called simply `Gravity`. I added a model called `Artwork` and then implemented CRUD endpoints without using Grape - just plain old Rails controllers. Then I installed Grape on the project and implemented those same CRUD operations using a Grape endpoint. I wrote request specs to verify that the two sets of endpoints behave the same way.

Then I created [a PR that implements something new](https://github.com/jonallured/pretend_gravity/pull/1) - adds a new field to the `Artwork` model that should not be allowed to be set via the API. I used this PR to demonstrate that it's too easy for a Grape user to write endpoint code that is not protected from Mass Assignment abuse. [The last commit on that PR](https://github.com/jonallured/pretend_gravity/pull/1/commits/d9d67b1e3ddc645637913376b3259467b56811c9) adds a failing test that shows the abuse in action. That same type of test was written on the Rails side and passes because over there we have protection via Strong Params from Rails.

Then I wrote a series of mitigations that might be attempted.

## Mitigation One: use except

[The first mitigation](https://github.com/jonallured/pretend_gravity/commit/da9763a438c82949fa72f3fbfedb385581d390a4) I attempted was using [the `except` method](https://api.rubyonrails.org/classes/Hash.html#method-i-except) that removes a key from a hash. This passes the failing test but I'm not a fan because:

* every new field that's added would need to be removed
* an engineer would have to know that they have to use this method to keep their params safe

## Mitigation Two: use declared

[The second mitigation](https://github.com/jonallured/pretend_gravity/commit/16581c9a932f7d92288de88357db4862856f3173) uses [a helper from Grape called `declared`](https://github.com/ruby-grape/grape?tab=readme-ov-file#declared). There were two steps here:

* specify the params for the create and update endpoints
* wrap the `params` with the `declared` method and specify `include_missing` as `false`

This approach filters out params that are unknown and passes our failing tests. Even better it's future proof in this endpoint so that should a field be added in the future then it does not have to be manually added to the list of exceptions like in the first mitigation.

Still, as in the first mitigation, an engineer would have to know that they should use `declared` and future endpoints have no build-in protections. This one is better than the last mitigation but imho is still insufficient.

## Mitigation Three: use Strong Params

[The third mitigation](https://github.com/jonallured/pretend_gravity/commit/2a073084a1f84061b401ec6582bc59b2d49f3952) brings patterns from Rails into Grape by using [Strong Params](https://guides.rubyonrails.org/action_controller_overview.html#strong-parameters) and an `artwork_params` helper to pass the failing specs.

Failing tests are back to passing but we still have only addressed this particular endpoint and haven't done enough to protect our future selves from making similar mistakes.

## Mitigation Four: set Strong Params as param_builder

[The fourth mitigation](https://github.com/jonallured/pretend_gravity/commit/10682897a9a4e290c14faebfccba266720aa3360) takes the idea of using Strong Params a step further. Using [the `param_builder` configuration option](https://github.com/ruby-grape/grape?tab=readme-ov-file#params-class) of Grape I created a shim module that turns all `params` objects into `ActionController::Parameters` objects. I still had to add a helper for `artwork_params` and call `permit` with the allowed params as both the third mitigation and Rails controller must.

What seems better here is that our current endpoint and all future endpoints now benefit from the built-in protection offered by Strong Params. It's not _just_ the filtering of params, it's _also_ that if used incorrectly an exception is raised:

```
params = ActionController::Parameters.new({})
artwork = Artwork.create(params)
# => ActiveModel::ForbiddenAttributesError raised!
```

## Proposal for Adding Strong Params to Grape

So this was all in service of this: I propose we add Strong Params to Grape. There are already 3 options so there are patterns we could follow. Docs could be updated and should probably have a section about Mass Assignment and the stance of the Grape project. I think this would be a great first step in improving protections for Mass Assignment abuse.

I would take it further though. I would propose that eventually Strong Params should be the default and then even remove the other options altogether. Maybe it happens over the course of a few major releases to ease the upgrade pain but I believe it should happen. Rails had to go through this pain but now lots of apps are much more secure than they were before and I think Grape should have this reckoning too.

## Am I Missing Something?

I've looked at this every which way I could think of but maybe I'm missing something - are there other mitigations that I'm not thinking of? Is there a better way to address Mass Assignment abuse in Grape? How would you tackle this one?

And it's possible that adding Strong Params to Grape has already been discussed - let me know if there's something I can catch up on! Ultimately what I'm after is better Mass Assignment protection in Grape. I believe Strong Params is the right way to address this but def open to hearing more.

**Comment by dchandekstark** [comment]
Based on my level of understanding of this issue, I am not convinced that Strong Params is the best answer for Grape, for these reasons:
- The ability to enforce `declared` globally on params could be explored further.
- Strong Params would add another gem dependency, which could require more frequent patching of Grape.
- Some of the benefits depend on using Rails models (AR/AM).  
- Some benefits for Rails models could be achieved by implementing `permitted?` on `params` (`false`) and the hash returned by `declared` (`true`).

**Comment by dblock** [comment]
@dchandekstark We added contract support in https://github.com/ruby-grape/grape/issues/2386. Is that a workable alternative?

**Comment by dchandekstark** [comment]
@dblock I hadn't seen that, but it looks promising, and is perhaps another point against the current proposal -- at least insofar as it wants tp replace current features with strong params.

**Comment by dblock** [comment]
@dchandekstark we haven't shipped the contract support yet (2.1.0), will you give it a try? would love your opinion

**Comment by dchandekstark** [comment]
> @dchandekstark we haven't shipped the contract support yet (2.1.0), will you give it a try? would love your opinion

Yes, I would like to try the contract approach, mainly due to the nicer encapsulation as compared to the default params.  I am wondering if this statement in the doc is accurate though, because, if so, I don't understand it (I am not that familiary dry-schema, however): 

"The latter will define a coercing schema (Dry::Schema.Params). When using the former approach, it's up to you to decide whether the input will need coercing."
